lib/IDE/TypeContextInfo.cpp

```diff
@@ -85,19 +87,42 @@ void ContextInfoCallbacks::completeCaseStmtBeginning(CodeCompletionExpr *E) {
   // TODO: Implement?
 }
 
+class TypeContextInfoCallback : public TypeCheckCompletionCallback {
+  Expr *ParsedExpr;
+  SmallVector<Type, 2> Types;
+
+  void sawSolutionImpl(const constraints::Solution &S) override {
+    if (!S.hasType(ParsedExpr)) {
+      return;
+    }
+    if (Type T = getTypeForCompletion(S, ParsedExpr)) {
+      Types.push_back(T);
+    }
+  }
+
+public:
+  TypeContextInfoCallback(Expr *ParsedExpr) : ParsedExpr(ParsedExpr) {}
+
+  ArrayRef<Type> getTypes() const { return Types; }
+};
+
 void ContextInfoCallbacks::doneParsing(SourceFile *SrcFile) {
   if (!ParsedExpr)
     return;
 
-  typeCheckContextAt(TypeCheckASTNodeAtLocContext::declContext(CurDeclContext),
-                     ParsedExpr->getLoc());
-
-  ExprContextInfo Info(CurDeclContext, ParsedExpr);
```

Comment:
Could we remove `#include "ExprContextAnalysis.h"`?

Reply:
We still need it for
```cpp
void typeCheckContextAt(TypeCheckASTNodeAtLocContext TypeCheckCtx,
                        SourceLoc Loc);
```

At some point, we should probably just move that function somewhere else.